llvm/include/llvm/CodeGen/RegAllocEvictionAdvisor.h

```diff
@@ -191,13 +193,66 @@ class RegAllocEvictionAdvisorAnalysis : public ImmutablePass {
   const AdvisorMode Mode;
 };
 
+/// Common provider for legacy and new pass managers.
+/// This keeps the state for logging, and sets up and holds the provider.
+/// The legacy pass itself used to keep the logging state and provider,
+/// so this extraction helps the NPM analysis to reuse the logic.
+class RegAllocEvictionAdvisorProvider {
+public:
+  enum class AdvisorMode : int { Default, Release, Development };
+  RegAllocEvictionAdvisorProvider(AdvisorMode Mode) : Mode(Mode) {}
+
+  virtual ~RegAllocEvictionAdvisorProvider() = default;
+
+  virtual bool doInitialization(Module &M) { return false; }
+
+  virtual void logRewardIfNeeded(const MachineFunction &MF,
+                                 llvm::function_ref<float()> GetReward) {}
+
+  virtual std::unique_ptr<RegAllocEvictionAdvisor>
+  getAdvisor(const MachineFunction &MF, const RAGreedy &RA) = 0;
+
+  /// Set the analyses that the advisor needs to use as they might not be
+  /// available before the advisor is created.
+  virtual void setAnalyses(std::initializer_list<llvm::Any> AnalysisP) {}
+
+  AdvisorMode getAdvisorMode() const { return Mode; }
+
+private:
+  const AdvisorMode Mode;
+};
+
+RegAllocEvictionAdvisorProvider *createReleaseModeAdvisorProvider();
```

Comment:
do we need these `create` methods? Are they just an artifact that's not applicable anymore, coming from the LPM conversion?

Reply:
In fact, even more - I think you can drop the `doInitialization` from the providers and just pass the module in the ctor, since - iiuc - the `create...` functions are called in places you have a `Module`. In the LPM analyses' case, you can model the fields as `std::unique_ptr`s and dynamically initialze them there, wdyt?

Reply:
I'll try that.

Reply:
Can drop `doInitialization`, not sure how to work around declaring the `create..Advisor()`'s

The ML and Release mode providers are in `MLRegAlloc..cpp` so I wanted to expose only the "base" signature to create those providers in this default impl file `RegAllocEvictionAdvisor.cpp`

Reply:
Can't the respective analyses just instantiate the appropriate provider at their construction?

Reply:
They do for the legacy analyses, but these are only used for the new (and only one) analysis.
Without the methods, the new PM analysis has to be in `MLRegAlloc..cpp` to keep the ML advisors local there, and also have to move the Default Provider out to `llvm::`.
Or - just separate out the ML advisor initialization in the new analysis to the `MLRegAlloc..cpp`?

Reply:
> Or - just separate out the ML advisor initialization in the new analysis to the `MLRegAlloc..cpp`?

Yes, each analysis would instantiate its provider as its implementation detail.

Reply:
Removed them, separated out the ML initialization in the new PM analysis class to the ML implementation file.